rskj-core/src/main/java/co/rsk/mine/gas/provider/MinGasPriceProviderFactory.java

```diff
@@ -14,7 +15,7 @@ public class MinGasPriceProviderFactory {
     private MinGasPriceProviderFactory() {
     }
 
-    public static MinGasPriceProvider create(long fixedMinGasPrice, StableMinGasPriceSystemConfig stableMinGasPriceSystemConfig) {
+    public static MinGasPriceProvider create(long fixedMinGasPrice, StableMinGasPriceSystemConfig stableMinGasPriceSystemConfig, OnChainMinGasPriceProvider.GetContextCallback getContextCallback) {
```

Comment:
When you pass a callback, that callback is only ever called once the ethmodule exists, as there are no logical paths to it otherwise. When you pass the output of the getEthModule function you then have a runtime circular dependency ie halting problem, because at the time of instantiation of the miner-related objects ethModule does not exist. And to instantiate ethModule you need to pass those miner-related objects.

Reply:
But I agree that this is not very "Java-esque". My JS background shows, and can't think of other non-disruptive way to achieve this.